src/main/java/net/sf/mzmine/modules/peaklistmethods/identification/ms2search/Ms2SearchTask.java

```diff
@@ -0,0 +1,280 @@
+     */
+    public double getFinishedPercentage() {
+        if (totalRows == 0)
+            return 0;
+        return ((double) finishedRows) / totalRows;
+    }
+
+    /**
+     * @see net.sf.mzmine.taskcontrol.Task#getTaskDescription()
+     */
+    public String getTaskDescription() {
+        return "MS2 similarity comparison between " + peakList1 + "and"
+                + peakList2;
+    }
+
+    /**
+     * @see java.lang.Runnable#run()
+     */
+    public void run() {
+
+        setStatus(TaskStatus.PROCESSING);
+
+        logger.info("Starting MS2 similarity search between " + peakList1 + "and"
+                + peakList2 + "with mz tolerance:"+mzTolerance.getPpmTolerance());
+
+        Ms2SearchResult searchResult;
+        double score;
+        int numIonsMatched;
+        PeakListRow rows1[] = peakList1.getRows();
+        PeakListRow rows2[] = peakList2.getRows();
+        
+        int rows1Length = rows1.length;
+        int rows2Length = rows2.length;
+        
+        totalRows = rows1Length;
+        
+        for (int i = 0; i < rows1Length; i++) {
+            for (int j = 0; j < rows2Length; j++) {
+                Feature featureA = rows1[i].getBestPeak();
+                Feature featureB = rows2[j].getBestPeak();
+                
+                int featureAID = featureA.hashCode();
+                int featureBID = featureB.hashCode();
+                
+                searchResult = simpleMS2similarity(featureA,
+                        featureB, intensityThreshold, mzTolerance.getPpmTolerance());
+                
+                //Report the final score to the peaklist identity
+                if (searchResult != null && searchResult.getScore() > scoreThreshold && searchResult.getNumIonsMatched() >= minimumIonsMatched)
+                    addFragmentClusterIdentity(rows1[i],featureA,featureB,searchResult);
+                
+                //featureA.appendFragmentSimilarityScore(featureBID,result);
+              
+                if (isCanceled())
+                    return;
+            }
+            
+        //Update progress bar
+        finishedRows++;
+        }
+
+        // Add task description to peakList
+        ((SimplePeakList) peakList1).addDescriptionOfAppliedTask(
+                new SimplePeakListAppliedMethod("Identification of complexes",
+                        parameters));
+
+        // Repaint the window to reflect the change in the peak list
+        Desktop desktop = MZmineCore.getDesktop();
+        if (!(desktop instanceof HeadLessDesktop))
+            desktop.getMainWindow().repaint();
+
+        setStatus(TaskStatus.FINISHED);
+
+        logger.info("Finished MS2 similarity search for " + peakList1 + "against"
+                + peakList2);
+
+    }
+
+    /**
+     * Simple cosine like calculation of similarity between ions of two fragmentation scans.
+     * 
+     */
+    private Ms2SearchResult simpleMS2similarity(Feature featureA, Feature featureB,
```

Comment:
I would suggest to move this function into a separate class (e.g. CosineProductSimilarity) and move it to the util package. Then we can use the function here as well as in the scatter plot, for example.